File: lib/models/facility.dart
enum FacilityType {
  hospital,
  clinic,
  shelter,
  petCare,
  medicineProvider,
}

extension FacilityTypeExtension on FacilityType {
  String get displayName {
    switch (this) {
      case FacilityType.hospital:
        return 'Veterinary Hospital';
      case FacilityType.clinic:
        return 'Pet Clinic';
      case FacilityType.shelter:
        return 'Animal Shelter';
      case FacilityType.petCare:
        return 'Pet Care Provider';
      case FacilityType.medicineProvider:
        return 'Medicine Provider';
    }
  }
}

class Facility {
  final String id;
  final String name;
  final FacilityType type;
  final bool isVerified;
  final String imageUrl;
  final String address;
  final String city;
  final double distanceKm;
  final String phone;
  final double rating;
  final String openHours;
  final bool isOpenNow;
  final bool hasEmergency;
  final bool offersSupportedCare;
  final List<String> services;
  final int totalAnimalsHelped;
  final int availableCapacity;

  Facility({
    required this.id,
    required this.name,
    required this.type,
    required this.isVerified,
    required this.imageUrl,
    required this.address,
    required this.city,
    required this.distanceKm,
    required this.phone,
    required this.rating,
    required this.openHours,
    required this.isOpenNow,
    required this.hasEmergency,
    this.offersSupportedCare = true,
    required this.services,
    this.totalAnimalsHelped = 0,
    this.availableCapacity = 0,
  });
}

File: animal-connect-backend/src/models/Facility.js
const mongoose = require('mongoose');

const facilitySchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: [true, 'Facility name is required'],
      trim: true
    },
    type: {
      type: String,
      enum: ['hospital', 'clinic', 'shelter', 'petCare', 'medicineProvider'],
      required: [true, 'Facility type is required']
    },
    owner: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    email: {
      type: String,
      trim: true,
      lowercase: true
    },
    phone: {
      type: String,
      required: [true, 'Contact phone is required'],
      trim: true
    },
    address: {
      type: String,
      required: [true, 'Address is required'],
      trim: true
    },
    city: {
      type: String,
      required: [true, 'City is required'],
      trim: true
    },
    location: {
      type: {
        type: String,
        enum: ['Point'],
        default: 'Point'
      },
      coordinates: {
        type: [Number], // [longitude, latitude]
        required: true,
        default: [76.9558, 11.0168] // Default coords (e.g. Coimbatore center)
      }
    },
    isVerified: {
      type: Boolean,
      default: false
    },
    verificationStatus: {
      type: String,
      enum: ['pending', 'verified', 'rejected', 'suspended'],
      default: 'pending'
    },
    verificationRemarks: {
      type: String,
      default: ''
    },
    verifiedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    verifiedAt: {
      type: Date
    },
    imageUrl: {
      type: String,
      default: 'https://images.unsplash.com/photo-1583337130417-3346a1be7dee?w=600'
    },
    rating: {
      type: Number,
      default: 4.8,
      min: 0,
      max: 5
    },
    openHours: {
      type: String,
      default: '24 Hours Emergency'
    },
    isOpenNow: {
      type: Boolean,
      default: true
    },
    hasEmergency: {
      type: Boolean,
      default: true
    },
    offersSupportedCare: {
      type: Boolean,
      default: true
    },
    services: [
      {
        type: String
      }
    ],
    totalAnimalsHelped: {
      type: Number,
      default: 0
    },
    capacity: {
      type: Number,
      default: 0
    },
    currentOccupancy: {
      type: Number,
      default: 0
    },
    registrationDocuments: [
      {
        title: { type: String, default: 'Registration Certificate' },
        url: { type: String, required: true },
        uploadedAt: { type: Date, default: Date.now }
      }
    ]
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true }
  }
);

// Geospatial index for nearby queries
facilitySchema.index({ location: '2dsphere' });
facilitySchema.index({ type: 1, isVerified: 1 });

// Virtual for available capacity
facilitySchema.virtual('availableCapacity').get(function () {
  if (this.type === 'shelter') {
    return Math.max(0, this.capacity - this.currentOccupancy);
  }
  return 0;
});

const Facility = mongoose.model('Facility', facilitySchema);
module.exports = Facility;
